### ISSUE TITLE
Use GNU getopts?

### ISSUE THREAD

**Issue by tavinus** [willemdh/FireMotD#5]
I believe this script is Linux only right? (no BSD or Mac or anything else?).

Because I was wondering if it would be OK to use GNU getopts to parse the options/parameters. It is quite easy to implement and quite powerful to use. We would have many benefits, like being able to use the parameters in any order, better parameter validation, etc.

The only downside is that GNU getopts is not available in other *nixes than Linux (by default, but can be installed). I guess the real downside is to create a dependency on GNU getopts, but since this is Linux only and all linuxes have GNU getopts...

I know for sure this doesn't work on Macs (I tried), and all other BSDs are probably the same.
Of course, if you plan to expand this to other systems, this will require extra work. But I think it would be nicer to just create another script for the other systems (keep it small, keep it clean).

If it is OK for you I will implement parsing the options/parameters with GNU getopts and make a pull request.

**EDIT:** After some nightmare conflicts I was able to rebase my fork to yours and have already made some minor changes. But I will wait for your answer to make another pull request. I am still not sure on why I couldn't merge with upstream, since you seem to have pulled all my changes and then changed on top of them. Did you edit the file on the browser by any chance?

Cheers!
Gus

**Comment by willemdh** [comment]
@tavinus I'm not sure about this. I really like the way I'm using the parameters with the

    while :; do
    case "$1" in

Not sure what you mean with benefits such as being able to use the parameters in any order, better parameter validation? 

Another small note is that when I find some time I would like to port it to Bash Windows (http://www.pcworld.com/article/3049664/windows/hell-freezes-over-microsoft-partners-with-ubuntu-to-bring-bash-shell-to-windows-10.html)

I've used getopts in the past, but I thought my method was cleaner. Could you maybe post a short example here as how you would do it? TX!

**Comment by tavinus** [comment]
I see...
That is kind of why I asked. Some times there are portability problems and some times there are just preference.

Better validation:
- Nowadays the script doesn't care if one puts rubbish after a valid command
- Verbose -v ONLY works if it comes before any other command
- The script probably shouldn't be able to run both -t theme and -C "text" at the same time, even though this could be fixed without getopt and is not critical.

When porting to Windows or even BSDs the script will need to run very differently on each system to collect information. Just because the programs used to collect system data will be completely different ones.

Now, BSDs have their own getopt, BASH has its own built-in getopt as well (only short names I think). If you instal cygwin on windows you get GNU getopt as well. But in any case, porting from GNU's getopt to BSD/Apple one should be trivial as well. Or you can even use the way it is now (without getopt) on those systems.

But I kind of think it is a bad idea to keep all systems in ONE script. It would be much better just to have some other file, say: `WinFireMotD` or whatever you want. You may even think on having a main FireMotD script that detects the system and calls an auxiliary script designed for that system.

I have been using GNU getopts on some scripts and I like it. Here is an example from a script I have in production. It parses options with arguments and parameters.

```bash
# read the options
TEMP=`getopt -o vqhV --long verbose,quiet,version,help -n '/usr/bin/mixsync' -- "$@"`
eval set -- "$TEMP"
#echo "TEMP: $TEMP"

PARAMS=''

# extract options and their arguments into variables.
while true ; do
	case "$1" in
		-v|--verbose)
			VERBOSE=$TRUE
			PARAMS='-v'
			shift ;;
		-q|--quiet)
			VERBOSE=$FALSE
			PARAMS='-q'
			shift ;;
		-h|--help)
			print_help
			exit 0 ;;
		-V|--version)
			print_version
			exit 0 ;;
		--) 
			shift ; 
			break ;;
		*)
			usage
			exit 1 ;;
    esac
done
```
This is up to what would be used on FireMotD. We would be done.

Bur after that `$1` points to the 1st parameter (which has no - or --). These parameters are orders for my program on this case, which call a GPS tracking webservice using PHP. I am using if statements, but if could be a case statement as well.
```bash
if [[ $1 = 'positions' ]]; then
	$PHP_BIN $MIX_SBIN_PATH/$MIX_POSITIONS $PARAMS
	exit $?
elif [[ $1 = 'vehicles' ]]; then
	$PHP_BIN $MIX_SBIN_PATH/$MIX_VEHICLES $PARAMS
	exit $?
elif [[ $1 = 'locations' ]]; then
	$PHP_BIN $MIX_SBIN_PATH/$MIX_LOCATIONS $PARAMS
	exit $?
elif [[ $1 = 'eventsDescriptions' ]]; then
	$PHP_BIN $MIX_SBIN_PATH/$MIX_EVENTS_DESCRIPTIONS $PARAMS
	exit $?
elif [[ $1 = 'sendRelayCommand' ]]; then
	if [[ (-z $2) || (-z $3) ]]; then
		echo $'\n'"Numero de parametros insuficiente!";
		echo " -> mixsync sendRelayCommand <idVeiculo> <On|Off|Released|Progressive> [numeroRelay=1]"$'\n';
		usage
		exit 1;
	fi
	relay_number="1"
	[[ ! -z $4 ]] && relay_number="$4"
	#echo "RelayNumber: -> $relay_number - 4: $4"
	$PHP_BIN $MIX_SBIN_PATH/$MIX_SEND_RELAY_COMMAND $PARAMS $2 $3 $relay_number
	exit $?
elif [[ $1 = 'unlock' ]]; then
	$CASPER_BIN $MIX_SBIN_PATH/$MIX_UNLOCKER $REDIRECT
	exit $?
else
	echo "Erro:"
	echo "  Comando invalido:"
	echo "  -> $1"
	echo "Ajuda:"
	echo "  /usr/mixsync --help"
	exit 1;
fi
```
I hope the Portuguese there is not a problem :)
But anyways, we use sometimes like this:
`mixsync sendRelayCommand 2 On -v`
`mixsync sendRelayCommand 2 On 1 -v`

Note that I can put any parameter anywhere, but of course I have to put the sendRelayCommand parameters after it on the right order (even though it has a last parameter that is optional). After we run getopt it canonizes it for us.

Well.. anyways...
Just leave it be for now..
But I have a pull request for you with many changes and fixes anyways.

Cheers!
Gus

**Comment by tavinus** [comment]
Just a note:
On the article he says 
> This is a genuine Ubuntu image on top of Windows with all the Linux tools I use like awk, sed, grep, vi, etc.,
> 

It is very much probable that GNU getopt is there then. Since many Linux tools use it.

**Comment by willemdh** [comment]
Well the way you describe it is actually the way I did parameter handling a few years ago. I just thought that my scripts were cleaner with the way I'm handling it now. (without getop  or getopt)
When I make a change I only need to add a parameter in one line and not in the TEMP=`getopt line. 

It's an interesting discussion though. By the way I have a lot more ideas for the FireMotD. You probably have noticed some pieces in it that should eventually lead to it. On my to do are:
- Finishing the html table output
- Converting the updatecount text file to a general file containing more information then only the update count. For example the apache, php version etc takes too long to load, so I will put it in a cron job too. 

If you like to help me with the above maybe we better continue talking on email. But I'm marrying next week, so I'm a but short on time atm. To be continued!!

**Comment by tavinus** [comment]
WOW! Congrats on your marriage! :)

Yeah, saving some slow stuff would be awesome. It takes up to 5 seconds to show on some machines I tested. 

Nevermind the getopt, it is fine the way it is. FireMotD does not need too many parameters that could be combined in different ways, so getopt is very optional, and eliminating one dependency is always nice.

I barely played with the html option, but I may have a look on it.

In bash it is probably easiest to have a text file to be read one line at a time. It is not a language very easy to tokenize though.

Did you have a look on the latest pull request I made.
Some stuff were kind of broken on the last version. Some stuff I changed.
Should be all fixed now.

Cheers!
And Congrats for the Marriage!

Gus
